docs/proposals/20221611-workload-identity-integration.md

```diff
@@ -0,0 +1,351 @@
+
+### <a name='Goals'></a>Goals
+
+- Integrate Azure Workload Identity into capz.
+
+### <a name='Non-GoalsFutureWork'></a>Non-Goals / Future Work
+
+- Migration to `azidentity` from `adal` module when using `go-autorest` library in capz as `adal` will be deprecated. 
+
+## <a name='Proposal'></a>Proposal
+
+In this model, Kubernetes cluster itself becomes the token issuer issuing tokens to Kubernetes service accounts. These service accounts can be configured to be trusted on Azure AD application or user assigned identity. Workloads/pods can use this service account token which is projected to it's volume and can exchange for an Azure AD token.
+
+The first step for creating K8s cluster using capz is creating a bootstrap cluster which is a Kind cluster and then deploying capi, capz and other helper components. On a high level the workflow looks the following to be able to use AWI and create a management cluster.
+
+**Bootstrap Cluster**
+
+- The operator/admin should be able generate signing key pair or BYO key pair. 
+
+- The bootstrap cluster i.e the Kind K8s cluster should be configured with appropriate flags on kube-apiserver and kube-controller-manager and the key pairs should be mounted on the container path for control plane node. 
+
+- The operator/admin should set up OIDC issuer URL which means having access to azure blob storage where the following two documents will be stored in the blob storage container.
+  - Generate and upload the Discovery Document.
+  - Generate and upload the JWKS Document.
+ 
+- The AWI mutating admission webhook should be installed on the bootstrap cluster.
+
+- A service account should be created in the same namespace as that of capz pod and tagged with the client ID of the Azure AD application or user addigned identity on the bootstrap cluster. The service account should also be associated as a Subject with the appropriate ClusterRoleBindings and RoleBindings for the usual RBAC. This is the service account that the capz pod should use.
+
+- The capz manager service account should be annotated with the Azure AD or User Assigned Client ID. 
+
+- A fedreated credential should be created between the identity and the OIDC issuer URL and the service account(subject). 
+
+- CAPZ pod should use the `azidentity` module from the azure-sdk-for-go to exchange AD token in lieu of the projected service account token.
+
+**Management Cluster**
+
+- Once a K8s cluster is created from the bootstrap cluster, to convert it into a management cluster `clusterctl init` and `clusterctl move` commands are executed.
+
+- After `clusterctl init` command is executed, the control plane node should have the key pairs and then patched to include them in the container path and set the kube-apiserver and kube-controller-manager flags.
+
+- The capz pod service account should be annotated with the client ID of the User Assigned Identity. The client ID can be passed as an argument to the `clusterctl init` command. 
```

Comment:
The client ID annotation is **not required**. It's optional and can be omitted. In this scenario, CAPZ controller can use the clientID information from `AzureClusterIdentity`.

Reply:
So what I meant here is the following
- Most oftenly, a Kind cluster is used for bootstrap cluster for simplicity ( though not a requirement ). Or at least I have seen this workflow to users/customers whom I have interacted with. 
- Now, when a bootstrap cluster is created cluster and all the pre-requisites are met including annotating the service account. If not annotating then in some way client ID should be provided.
- After this, a workload cluster is created on azure. This workload cluster now needs to be converted to a management cluster. This means it should also meet all the requirements including a way to provide client-id that should be used. 
- Now, passing client-id via clusterctl was just a naive thought from me to somehow pass the client ID to the management cluster. 
- But as @aramase said client ID is optional ( I misunderstood it as required ) 

So I think specifying which client ID to use should not be complex and I will remove this passing client ID via clusterctl part. 
Also on the other had, I think we should now refrain from using AzureClusterIdentity and may be introducing a client ID field itself in the AzureCluster.

Reply:
Though now I am thinking if we can use AzureClusterIdentity if it provides more value then just helping pass a information which is client ID. 
Let us discuss on this.

Reply:
I think for back compat reasons it might be easier to keep AzureClusterIdentity for now so existing clusters can migrate. Later we can think about deprecating AzureClusterIdentity and just having a clientID field on AzureCluster